src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,320 @@
+              waitExecutionBackoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get().execute(request);
+
+        // We don't want to reset backoff for Execute() since if there is an error:
+        //   1. If happened before we received a first response, we want to ensure the retry counter
+        //      is increased and call Execute() again.
+        //   2. Otherwise, we will fallback to WaitExecution() loop.
+        //
+        // This also prevent us from being stuck at a infinite loop:
+        //   Execute() -> WaitExecution() -> Execute()
+        return handleOperationStream(operationStream, /* resetWaitExecutionBackoff */ false);
+      } catch (StatusRuntimeException e) {
+        if (lastOperation != null) {
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleOperationStream(operationStream, /* resetWaitExecutionBackoff */ true);
+      } catch (StatusRuntimeException e) {
+        // A NOT_FOUND error means Operation was lost on the server, retry Execute().
+        //
+        // However, we only retry Execute() if executeBackoff should retry. Also increase the retry
+        // counter at the same time since we will technically retry (done by nextDelayMillis()).
+        if (e.getStatus().getCode() == Code.NOT_FOUND && executeBackoff.nextDelayMillis() >= 0) {
+          lastOperation = null;
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    /** Process a stream of operations from Execute() or WaitExecution(). */
+    ExecuteResponse handleOperationStream(Iterator<Operation> operationStream,
+        boolean resetWaitExecutionBackoff) throws IOException {
+      try {
+        while (operationStream.hasNext()) {
+          Operation operation = operationStream.next();
+          ExecuteResponse response = handleOperation(operation);
+
+          // At this point, we received the response without error.
+          lastOperation = operation;
+
+          if (resetWaitExecutionBackoff) {
```

Comment:
What harm would it do to unconditionally reset this backoff? When called from Execute(), the Execute() backoff is bumped, so we don't have an infinite loop.

Reply:
Good catch! I cannot come up with a case where this could be an issue. Changed.